torch/csrc/jit/fuser/codegen.cpp

```diff
@@ -77,10 +81,54 @@ static const char* scalarTypeName(const at::ScalarType type) {
   }
 }
 
+static const char* calcScalarTypeName(const at::ScalarType type) {
+  if (type == at::ScalarType::Half) {
+    return "float";
+  }
+  return scalarTypeName(type);
+}
+
+
+static std::string variableType(const std::shared_ptr<c10::Type> t) {
+  if (t->kind() == TypeKind::IntType) {
+    return "int";
+  } else if (t->kind() == TypeKind::FloatType) {
+    return "float";
+  } else if (t->kind() == TypeKind::TensorType) {
+    auto const tt = t->cast<TensorType>();
+    return calcScalarTypeName(tt->scalarType());
+  }
+  // TODO: we might also error out if we think the above should cover everything
```

Comment:
I think we really should assert that we can shape prop through fused kernels. They're very simple, and with the current code I can't really prove to myself that it's correct (mostly because of this `auto`).